bson/marshalvalue_cases_test.go

```diff
@@ -0,0 +1,127 @@
+// Copyright (C) MongoDB, Inc. 2023-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bson
+
+import (
+	"io"
+	"testing"
+
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/internal/assert"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+)
+
+// helper type for testing MarshalValue that implements io.Reader
+type marshalValueInterfaceInner struct {
+	Foo int
+}
+
+var _ io.Reader = marshalValueInterfaceInner{}
+
+func (marshalValueInterfaceInner) Read([]byte) (int, error) {
+	return 0, nil
+}
+
+// helper type for testing MarshalValue that contains an interface
+type marshalValueInterfaceOuter struct {
+	Reader io.Reader
+}
+
+// helper type for testing MarshalValue that implements ValueMarshaler
+type marshalValueMarshaler struct {
+	Foo int
+}
+
+var _ ValueMarshaler = marshalValueMarshaler{}
+
+func (mvi marshalValueMarshaler) MarshalBSONValue() (bsontype.Type, []byte, error) {
+	return bsontype.Int32, bsoncore.AppendInt32(nil, int32(mvi.Foo)), nil
+}
+
+var _ ValueUnmarshaler = &marshalValueMarshaler{}
+
+func (mvi *marshalValueMarshaler) UnmarshalBSONValue(t bsontype.Type, b []byte) error {
+	v, _, _ := bsoncore.ReadInt32(b)
+	mvi.Foo = int(v)
+	return nil
+}
+
+type marshalValueStruct struct {
+	Foo int
+}
+
+type marshalValueTestCase struct {
+	name     string
+	val      interface{}
+	bsontype bsontype.Type
+	bytes    []byte
+}
+
+func unmarshalValueTestCases(t *testing.T) []marshalValueTestCase {
+	t.Helper()
+
+	oid := primitive.NewObjectID()
+	regex := primitive.Regex{Pattern: "pattern", Options: "imx"}
+	dbPointer := primitive.DBPointer{DB: "db", Pointer: primitive.NewObjectID()}
+	codeWithScope := primitive.CodeWithScope{Code: "code", Scope: D{{"a", "b"}}}
+	idx, scopeCore := bsoncore.AppendDocumentStart(nil)
+	scopeCore = bsoncore.AppendStringElement(scopeCore, "a", "b")
+	scopeCore, _ = bsoncore.AppendDocumentEnd(scopeCore, idx)
+	decimal128 := primitive.NewDecimal128(5, 10)
+	structTest := marshalValueStruct{Foo: 10}
+	structCore, err := Marshal(structTest)
```

Comment:
Wrapping this data in `var()` would align the "=" values and make this more readable.